design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+
+## Test plan
+
+Aside from the usual unit tests etc, the QTT suit of tests will be extended to cover the different
+key formats. Tests will be added to cover the new syntax and configuration combinations and 
+permutations. Existing tests covering aggregations, re-partitions and joins will be extended to 
+include variants with different key formats. 
+
+## LOEs and Delivery Milestones
+
+The KLIP will be broken down into the following deliverables:
+
+1. **Basic JSON support**: Support for the `JSON` key format, without:
+    * schema registry integration
+    * making KAFKA_TOPIC property optional
+    * Automatic repartitioning of streams and tables for joins where key formats do not match: such
+      joins will result in an error initially.
+  
+    Included in this milestone:
+    
+    * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
+    * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
+    * Addition of new server configuration to provide defaults for key and value formats.
+    * Removal of requirement for `VALUE_FORMAT` property in the `WITH` clause of CT/CS statements, where
+      the server configuration provides a default.
+    * Support for additional key column data types, as JSON supports them:
+        * `DECIMAL`
+        * `BOOLEAN`
+    * Full support of these key formats for all supported SQL syntax  
+1. **Auto-repartitioning on key format mismatch**. Adds support for automatic repartitioning of 
+   streams and tables for joins where key formats do not match.
+1. **Schema Registry support**: Adds support for reading and writing schemas to and from the schema
+   registry.
+1. **JSON_SR support** Adds support for the `AVRO` key format, inc. schema registry integration.
+1. **Avro support** Adds support for the `AVRO` key format, inc. schema registry integration.
+1. **Delimited support**: Adds support for the `DELIMITED` key format.
+1. **Optional KAFKA_TOPIC property**: makes the `KAFKA_TOPIC` property optional everywhere and add
+   configuration to control case and case sensitivity of topic names.
+1. **Blog post**: write a blog post about the new features. (Potentially more than once if 
+  work span multiple releases).
+   
+LOE TBD once scope and design confirmed.
+
+## Documentation Updates
+
+New server config and new `CREATE` properties will be added to main docs site.
+
+There are no incompatible changes within the proposal, so no demos and examples _must_ change.
+However, it probably pays to update some to highlight the new features. We propose updating the 
+Kafka micro site examples to leverage the new functionality, as these have automated testing.  
+It may be worth changing the ksqlDB quickstart too - TBD, as this will require extending DataGen 
+to support other key formats. Something we may want in scope anyway - or should be end-of-life 
+DataGen in favour of the datagen connector?
```

Comment:
+1

Reply:
Please don't eol the datagen tool! 
It's far too useful for anyone who wants to test or demo something quickly. And before someone says "you can just run an embedded datagen connector though so what's the difference?" :-),  allow me to opine that this requires a steeper learning curve and advanced troubleshooting skills to get working right, which I think should be avoided where possible. Just an opinion of course ;)
I'd also say that the connector should be refactored (as i recall suggesting loudly when it was originally forked off from here) so that it embeds some re-usable portion of the datagen tool (which likely requires a small refactor of datagen itself too, to facilitate, to be fair) rather than being a copy/paste that now proceeds on its own life journey and inevitable divergence.

Reply:
@blueedgenick are you putting a case forward that DataGen should be enhanced to support Avro / Json keys as part of this work, i.e. it should be in-scaope?